test/gen_source.rb

```diff
@@ -0,0 +1,113 @@
+require 'optparse'
+require 'yaml'
+require 'erb'
+
+
+REGEXT_PRETTY = %r{
+ =\s            # We are interested to the right of the equal sign
+ (
+    ""|         # Empty string
+    ".*?[^\\]"| # String who can contain space and quoted string
+    [^\s,]+     # Anything except space and comma
+ )
+}x
+
+SOURCE_TEMPLATE =  <<-TEXT
+/* Code generaed by source_callbacks_generator.rb */
+
+#include "component.h"
+#include "create.h"
+
+void btx_push_usr_messages(struct common_data_s *common_data) {
+
+    <%- data.each do | entry | -%>
+    <%- entry.fetch(:times,1).times do -%>
+    btx_push_message_<%= entry[:name].gsub(":","_") %>(common_data,<%=  entry[:field_values].join(",") %>);
+    <%- end -%>
+    <%- end -%>
+}
+
+TEXT
+
+def parse_log(input_path)
+  data_list = []
+  File.open(input_path, "r") do |f|
+    f.each_line do |line|
+
+      # Line format support checks
+      match = line.match(REGEXT_PRETTY)
+      raise "Unsupported format for '#{line}'." unless match
+
+      i = line.index('{')
+      head, _, tail = line.partition(": {")
+      
+      data = {}
+      data[:name]=head.gsub(/[^0-9A-Za-z\-]/, '_') # Should reuse metababel mangling
+      data[:field_values] = tail.scan(REGEXT_PRETTY).flatten
+      data_list.append(data)
+    end
+  end
+  data_list
+end
+
+def parse_yaml(input_path)
+  data = YAML.load_file(input_path)
+  data.each do |item| 
+    item[:field_values] = item[:field_values].map(&:inspect)
+  end 
+end 
+
+def render_and_save(data, output_path) 
+  renderer = ERB.new(SOURCE_TEMPLATE, nil, '-')
+  output = renderer.result(binding)
+  File.write(output_path, output, mode: "w")
+end
+
+DOCS = <<-DOCS
+  Usage: example.rb [options]  
+  
+  Example: 
+    ruby example.rb -t yaml -i stream_classes.yaml -o callbacks.c
```

Comment:
So I don't understand why we need this option

Reply:
This generates source messages from the instances.yaml (bellow), we used this previously to generate the source push messages to test the tally. So at the end, we have two options: (1) generate source messages from yaml (below) or (2) do it from the babeltrace log.

```yaml
- :name: lttng:host
  # hostname, vpid, vtid, timestamp, backend ; name, duration, error
  :field_values: [iris01,1289,1098,1,0,host_function_name_1,2,0]
  :times: 5
...
```

Reply:
Oh, I see; if we go this road, we should have two codes.
`log -> yaml and yaml -> callbacks.c`